bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/test/java/com/google/cloud/bigtable/mirroring/hbase1_x/TestMirroringTableInputModification.java

```diff
@@ -0,0 +1,236 @@
+import com.google.common.util.concurrent.SettableFuture;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.Row;
+import org.apache.hadoop.hbase.client.Table;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentMatchers;
+import org.mockito.Mock;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+import org.mockito.stubbing.Answer;
+
+@RunWith(JUnit4.class)
+public class TestMirroringTableInputModification {
+  @Rule public final MockitoRule mockitoRule = MockitoJUnit.rule();
+
+  @Rule
+  public final ExecutorServiceRule executorServiceRule =
+      ExecutorServiceRule.singleThreadedExecutor();
+
+  @Mock Table primaryTable;
+  @Mock Table secondaryTable;
+  @Mock MismatchDetector mismatchDetector;
+  @Mock FlowController flowController;
+  @Mock SecondaryWriteErrorConsumerWithMetrics secondaryWriteErrorConsumer;
+
+  MirroringTable mirroringTable;
+  SettableFuture<Void> secondaryOperationAllowedFuture;
+
+  @Before
+  public void setUp() throws IOException, InterruptedException {
+    setupFlowControllerMock(flowController);
+    this.mirroringTable =
+        spy(
+            new MirroringTable(
+                primaryTable,
+                secondaryTable,
+                this.executorServiceRule.executorService,
+                mismatchDetector,
+                flowController,
+                secondaryWriteErrorConsumer,
+                new ReadSampler(100),
+                false,
+                new MirroringTracer()));
+
+    mockExistsAll(this.primaryTable);
+    mockGet(this.primaryTable);
+    mockBatch(this.primaryTable);
+
+    secondaryOperationAllowedFuture = SettableFuture.create();
+
+    blockMethodCall(secondaryTable, secondaryOperationAllowedFuture)
+        .existsAll(ArgumentMatchers.<Get>anyList());
+    blockMethodCall(this.secondaryTable, secondaryOperationAllowedFuture)
+        .batch(ArgumentMatchers.<Put>anyList(), (Object[]) any());
+    blockMethodCall(this.secondaryTable, secondaryOperationAllowedFuture)
+        .get(ArgumentMatchers.<Get>anyList());
+  }
+
+  @Test
+  public void testExistsAll() throws IOException {
+    List<Get> gets = createGets("k1", "k2", "k3");
+    List<Get> inputList = new ArrayList<>(gets);
+
+    this.mirroringTable.existsAll(inputList);
+    verify(this.primaryTable, times(1)).existsAll(inputList);
+    inputList.clear(); // User modifies the list
+
+    secondaryOperationAllowedFuture.set(null);
```

Comment:
Maybe rename it to communicate that secondary operation is blocked on this future. secondaryOperationBlockedOnFuture?

Reply:
Done.